### ISSUE TITLE
Enhancement: Clean up zero_to_fp32.py

### ISSUE THREAD

**Issue by siddk** [microsoft/DeepSpeed#1062]
Current `zero_to_fp32.py` script is in need of optimization; currently the existing implementation requires 2X RAM for the parameters themselves.

Additionally, add documentation for when this script is necessary, how to plug into HuggingFace.

cc @stas00.

**Comment by stas00** [comment]
Awesome! Thank you!

I'd start with adding a test first, so that the new version still works ;)

and the docs to plug into HF aren't needed. It's a built-in Deepspeed feature, it doesn't matter how uses it, as long as the checkpoint is saved it'll have this script bundled in the checkpoint folder.

**Comment by siddk** [comment]
Yup - happy to do so - will work through this in the next few days!

**Comment by stas00** [comment]
I don't recall when the script was added - I'd recommend working with the latest deepspeed both because transformers master requires it and there were bug fixes in deepspeed since 0.3.13.

**Comment by siddk** [comment]
@stas00 - happy to upgrade to newer DeepSpeed/Transformers.

But re: the note in the Transformer docs. We trained a GPT-2 model with HuggingFace 4.5.0 and DS 0.3.13, with ZeRO-2 (in FP16). Just spun up a notebook, and tried loading the checkpoint -- seems to say weights are all in FP32 by default?
![image](https://user-images.githubusercontent.com/2498509/117890427-ffed8b00-b269-11eb-96fc-4873f6e9303d.png)

**Comment by stas00** [comment]
Indeed, you get fp16 weights loaded as fp32 weights. Because the model is initialized in `dtype==fp32`, fp16 weights get auto-converted to `fp32`.

the easiest way to check that you're loading fp16 weights, other than manually doing `torch.load` is to look at the size of the file. under ds/fp16 it'll be half the size of the normal model.

and with `torch.load`:
```
$ python -c "import torch; sd = torch.load('output_dir/checkpoint-1/pytorch_model.bin'); print(sd['decoder.block.5.layer.2.DenseReluDense.wo.weight'].dtype)"
torch.float16
```

(this is from `t5-small`)

Hope this clarifies the confusion - and please let me know if it'd be good to further explain this in the docs and perhaps suggest how now that I trust I was able to explain what's going on behind the scenes.

**Comment by stas00** [comment]
The only way to get fp32 weights under deepspeed/fp16-mode is to reconsolidate those from the optimizer checkpoint files.

**Comment by siddk** [comment]
I think having this in the docs would be really good; it still isn't clear to me why you'd need to run the `zero_to_fp32.py` script if `.from_pretrained()` is enough. 

Especially for most usecases (fine-tuning, evaluating) involving models trained with DeepSpeed, that should be plenty?

**Comment by stas00** [comment]
the way deepspeed works is
1. stash fp32 master weights in optimizer
2. model.half()

so your `state_dict` is at fp16.

Regardless of zero2 or zero3 deepspeed's checkpoint optimizer always has fp32 master weights.

`zero_to_fp32.py` is there to recover those.

**Comment by stas00** [comment]
The only time you really want the fp32 weights if you release them on hub or share with someone - you're losing some precision when you downcast to fp16 weights.

So, you don't need `zero_to_fp32.py` for normal train/resume functioning. 

Definitely not if you are resuming from deepspeed checkpoint, since it has the fp32 weights there already.

**Comment by siddk** [comment]
And these optimizer master FP32 weights are always "more complete" than the `model.half()` weights? When other models are trained with FP16, do you have a sense as to whether they return the optimizer FP32 weights, or just the FP16 weights which they upcast to FP32?

**Comment by stas00** [comment]
ah, but no, there is the fp16 recovery under zero3 there as well, so it should be read.

**Comment by siddk** [comment]
Got it - I understand now, thank you so much for taking the time to explain this to me! Have you thought about just automating the save procedure with DeepSpeed to just run the export to FP32 under the hood? Presumably it doesn't take *that* long?

**Comment by stas00** [comment]
1. It takes a long time (unless your model is tiny), which can dramatically slow down the training - and why do that when the fp32 weights are already saved!
2. it needs a lot of memory, which the program already starved for it doesn't usually have - and it's very likely to OOM or run out of RAM.

**Comment by siddk** [comment]
Ok that's fair; alright well I have my work cut out for me - will start by writing some tests and going from there.

**Comment by stas00** [comment]
please note 2 critical gotchas 

1. I missed it in the first incarnation of this script - I used a model that had only one param_group and I was completely oblivious that it's not always the case. So the script failed as soon as multi-param-group model was run into. So for the test it's probably the best to create/use a small model that has 2-3 param groups and it will then work for 1-group too.
2. the data gets flattened into one huge flat array, so on reconstruction everything may appear to be of the right dimensions and loaded, but the data could be completely messed up.

**Comment by Mindful** [comment]
I debated opening a new issue for this but this seemed like the more appropriate place to mention it since you guys are already working on zero_to_fp32.py anyway.

At least for deepspeed 0.3.16, zero_to_fp32.py seems to drop content from state dicts that was registered with [register_buffer](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer). It looks like the current script gets states out of the optimizer, so maybe it would make sense that any state that isn't an optimizable model parameter gets dropped? The result is that running zero_to_fp32.py on models with buffers results in a new state dict that has fewer entries and consequently can't be used for loading.

I can try to create a minimal reproduction of this, but my hope is that the issue is straightforward enough that just mentioning it is sufficient. If a minimal reproduction is necessary just let me know and I can try and come back with one, although it may take some time.

**Comment by stas00** [comment]
@Mindful, since you're reporting is a bug/incomplete feature this warrants a dedicated issue, so that it's easier to track and resolve only when it's resolved. Thank you!

**Comment by piegu** [comment]
Hi @stas00. 

I have a question about how to get fp32 weight reconstruction. Your guide [Getting The Model Weights Out](https://huggingface.co/transformers/main_classes/deepspeed.html#getting-the-model-weights-out) is helpful in the case of `DeepSpeed` with a [HF script](https://github.com/huggingface/transformers/tree/master/examples/pytorch). Many thanks!

But what if `DeepSpeed` is used inside a [HF notebook](https://github.com/huggingface/notebooks/tree/master/examples)? In this case, we get the `trainer.model` after the training. Can we run the following code inside the notebook to get back the fp32 weigths?

`trainer.model.float();`

**Comment by stas00** [comment]
There is no difference between the script and the notebook model weights-wise.

- if you're using zero2 you can just get `trainer.model.float()` as you suggested, while losing some precision.

- with zero3 at the moment you either 
1. have to save a checkpoint and then restore fp32 weights as explained in the doc you linked to
2. and then you can call something like (untested):
```
pt_model = deepspeed_engine.module
state_dict = deepspeed_engine._zero3_consolidated_fp16_state_dict()
fp16_consolidated_model = pt_model.load_state_dict(state_dict)
```
and then to do the same as with zero2, and again you will lose some precision here.

May need to request for `_zero3_consolidated_fp16_state_dict` to be made public if someone has a direct use for it. Or could probably make the above into a wrapper that already returns the model.

To support you better it'd help if you could explain the use case.

**Comment by piegu** [comment]
Thanks @stas00. 

I read you agree with the following option in notebook for ZeRO-2:
```
trainer = Trainer(..., deepspeed='ds_config_zero2.json')
trainer.train()
trainer.model.float()
```

... but you add: _while losing some precision_

What would be another option (in order to not lose precision)? 
(in the case I want to save my model in the Model Hub of HF for example)

**Comment by stas00** [comment]
Use `zero_to_fp32.py` as explained in the doc you linked to.

**Comment by piegu** [comment]
Here is a notebook that shows a use case: [Fine-tuning a Masked Language Model (MLM) with DeepSpeed mixed-precision float (fp16)](https://colab.research.google.com/drive/17vN1w0HxRMopzps8WrlRwMmHhZppVQli?usp=sharing)

The idea is : how to use DeepSpeed with mixed-precision float (fp16) in the [Hugging Face notebooks](https://github.com/huggingface/notebooks/tree/master/examples).

I took the example of the [language-modeling.ipynb](https://github.com/huggingface/notebooks/blob/master/examples/language_modeling.ipynb). I modified the original notebook with DeepSpeed ZeRO-2 + fp16.

As expected, the `trainer.model` weights dtype is float16 after the training.

From the above messages, I understood it is better to use the `zero_to_fp32.py` script than running `trainer.model.float()` (... but I did not understand why).

Right?

I implemented this in the notebook and put a **WARNING** about the `trainer.push` method to the HF model hub as this method would push (in this case) fp16 model weights instead of fp32 ones. So here, you have to use the method [Use your terminal and git](https://huggingface.co/transformers/model_sharing.html#use-your-terminal-and-git).

Right?

**Comment by stas00** [comment]
> From the above messages, I understood it is better to use the `zero_to_fp32.py` script than running `trainer.model.float()` (... but I did not understand why).
> 
> Right?
> 
> I implemented this in the notebook and put a **WARNING** about the `trainer.push` method to the HF model hub as this method would push (in this case) fp16 model weights instead of fp32 ones. So here, you have to use the method [Use your terminal and git](https://huggingface.co/transformers/model_sharing.html#use-your-terminal-and-git).
> 
> Right?
> 
> **Note**: searching for the best model folder in order to run the `zero_to_fp32.py` script is not fun. Would be possible to integrate this script in the `finetuned-model.save_pretrained() `method?

**Comment by stas00** [comment]
> Here is a notebook that shows a use case: [Fine-tuning a Masked Language Model (MLM) with DeepSpeed mixed-precision float (fp16)](https://colab.research.google.com/drive/17vN1w0HxRMopzps8WrlRwMmHhZppVQli?usp=sharing)

This is very awesome, @piegu!

If you'd like when this is finished - you could contribute it to the HF notebook collection

> From the above messages, I understood it is better to use the `zero_to_fp32.py` script than running `trainer.model.float()` (... but I did not understand why).

It's because at the very last step, fp32 weights are converted to fp16 and some precision gets lost here. fp16 doesn't have all the numbers fp32 has, so some numbers get rounded to the nearest number fp16 has. Is this clear now?

> I implemented this in the notebook and put a **WARNING** about the `trainer.push` method to the HF model hub as this method would push (in this case) fp16 model weights instead of fp32 ones. So here, you have to use the method [Use your terminal and git](https://huggingface.co/transformers/model_sharing.html#use-your-terminal-and-git).

You could push fp16 weights to the hub too. I'm working on adding `dtype` to the config as we speak.

> **Note**: searching for the best model folder in order to run the `zero_to_fp32.py` script is not fun. Would be possible to integrate this script in the `finetuned-model.save_pretrained() `method?

Perhaps you could request a feature that prints out the path to the best checkpoint upon finishing training? 

But I hear you that it'd be ideal that all these things were to happen transparently. 

The main issue is that Deepspeed is primarily used with huge models which can't be easily loaded into memory on an already low-memory process, that's why the whole process was made separate.

Let me think on how to make it easy for small models. I understand the issue.

**Comment by stas00** [comment]
@piegu, I have tried to address your needs here: https://github.com/microsoft/DeepSpeed/pull/1181

Please try it and let me know what you think of it.

Thank you.
